### ISSUE TITLE
block-cipher: Block alignment

### ISSUE THREAD

**Issue by nickray** [RustCrypto/traits#159]
Some hardware implementations of AES need particular alignment of blocks to operate directly over AHB (e.g. NXP LPC55's HASHCRYPT wants 32-bit alignment, I can imagine this is a typical case for ARM microcontrollers). If a large chunk of data needs to be en/de-crypted that's aligned, it would be neat to pass this information through on the type level to avoid stack copies for each block. The `aligned` crate would be one approach to do it.

Splitting this out from https://github.com/RustCrypto/traits/issues/43 for possible discussion.

**Comment by tarcieri** [comment]
My suggestion on #43 was to have an associated `Block` type which can have a `#[repr(align(...))]`

**Comment by nickray** [comment]
Yeah, that might even be easier to parse mentally than having block generic over "`B`" :)

**Comment by tarcieri** [comment]
We use `generic-array` for other types like keys because there are real-world cases where they are larger than 32-bytes (e.g. AES-SIV), because that's a special case in pre-const generics Rust.

Blocks are a legitimate case where in practice they will probably never be larger than 32-bytes.

**Comment by newpavlov** [comment]
I am not sure if such change will be practical, at least with the current state of the Rust language. Don't forget that we build other APIs on top of `BlockCipher`. Such change would either require bubbling alignment restrictions to higher-level APIs and thus removing the convenience of using `&mut [u8]`, or we would have to keep two parallel sets of methods with and without those restrictions.

On modern x86 unaligned loads of aligned pointers have the same performance, but unfortunately it's indeed not true for ARM...
